go/libraries/doltcore/sqle/index/dolt_index.go

```diff
@@ -910,14 +929,132 @@ func (di *doltIndex) trimRangeCutValue(to int, keyPart interface{}) interface{}
 	return keyPart
 }
 
-func (di *doltIndex) prollyRangesFromSqlRanges(ctx context.Context, ns tree.NodeStore, ranges []sql.Range, tb *val.TupleBuilder) ([]prolly.Range, error) {
-	ranges, err := pruneEmptyRanges(ranges)
-	if err != nil {
+func (di *doltIndex) prollySpatialRanges(ctx context.Context, ns tree.NodeStore, ranges []sql.Range, tb *val.TupleBuilder) ([]prolly.Range, error) {
+	// should be exactly one range
+	rng := ranges[0][0]
+	lower, upper := sql.GetRangeCutKey(rng.LowerBound), sql.GetRangeCutKey(rng.UpperBound)
+
+	// TODO: no clue what this is for
+	tup := tb.BuildPermissive(sharePool)
+	if err := PutField(ctx, ns, tb, 0, lower); err != nil {
 		return nil, err
 	}
 
-	pranges := make([]prolly.Range, len(ranges))
+	minPoint, ok := lower.(sqltypes.Point)
+	if !ok {
+		panic("somehow received a not point in spatial index range")
+	}
+	maxPoint, ok := upper.(sqltypes.Point)
+	if !ok {
+		panic("somehow received a not point in spatial index range")
+	}
+
+	zMin := ZValue(minPoint)
+	zMax := ZValue(maxPoint)
+
+	pranges := make([]prolly.Range, 65)
+
+	// generate ranges for levels 0 - 31
+	for lvl := 0; lvl < 32; lvl++ {
+		minVal, maxVal := make([]byte, 17), make([]byte, 17)
+		minVal[0], maxVal[0] = byte(lvl), byte(lvl)
+		shamt := lvl << 1
+		zMin[1] = (zMin[1] >> shamt) << shamt
+		zMax[1] = (zMax[1] >> shamt) << shamt
+		binary.BigEndian.PutUint64(minVal[1:], zMin[0])
+		binary.BigEndian.PutUint64(minVal[9:], zMin[1])
+		binary.BigEndian.PutUint64(maxVal[1:], zMax[0])
+		binary.BigEndian.PutUint64(maxVal[9:], zMax[1])
```

Comment:
the encoding logic here is obscure and difficult to parse for most readers. move this logic to helper functions in `z_encoding.go`